src/dotnet/ExecutablePackageObtainer/ExecutablePackageObtainer.cs

```diff
@@ -0,0 +1,204 @@
+
+        private void InvokeRestore(
+            FilePath nugetconfig,
+            FilePath tempProjectPath,
+            DirectoryPath individualToolVersion)
+        {
+            _projectRestorer.Restore(tempProjectPath, individualToolVersion, nugetconfig);
+        }
+
+        private FilePath CreateTempProject(
+            string packageId,
+            PackageVersion packageVersion,
+            string targetframework,
+            DirectoryPath individualToolVersion)
+        {
+            FilePath tempProjectPath = _getTempProjectPath();
+            if (Path.GetExtension(tempProjectPath.Value) != "csproj")
+            {
+                tempProjectPath = new FilePath(Path.ChangeExtension(tempProjectPath.Value, "csproj"));
+            }
+
+            EnsureDirectoryExists(tempProjectPath.GetDirectoryPath());
+            if (packageVersion.IsConcreteValue)
+            {
+                File.WriteAllText(tempProjectPath.Value,
+                    string.Format(
+                        TemporaryProjectTemplate,
+                        targetframework,
+                        individualToolVersion.Value,
+                        packageId,
+                        packageVersion.Value));
+            }
+            else
+            {
+                File.WriteAllText(tempProjectPath.Value,
+                    string.Format(
+                        TemporaryProjectTemplateWithoutPackage,
+                        targetframework,
+                        individualToolVersion.Value));
+            }
+            return tempProjectPath;
+        }
+
+        private void InvokeAddPackageRestore(
+            FilePath nugetconfig,
+            FilePath tempProjectPath,
+            string packageId)
+        {
+            if (nugetconfig != null)
+            {
+                File.Copy(
+                    nugetconfig.Value,
+                    tempProjectPath
+                        .GetDirectoryPath()
+                        .CreateFilePathWithCombineFollowing("nuget.config")
+                        .Value);
+            }
+
+            _packageToProjectFileAdder.Add(tempProjectPath, packageId);
+        }
+
+        private DirectoryPath CreateIndividualToolVersionDirectory(
+            string packageId,
+            PackageVersion packageVersion)
+        {
+            DirectoryPath individualTool = _toolsPath.WithCombineFollowing(packageId);
+            DirectoryPath individualToolVersion = individualTool.WithCombineFollowing(packageVersion.Value);
+            EnsureDirectoryExists(individualToolVersion);
+            return individualToolVersion;
+        }
+
+        private static void EnsureDirectoryExists(DirectoryPath path)
+        {
+            if (!Directory.Exists(path.Value))
+            {
+                Directory.CreateDirectory(path.Value);
+            }
+        }
+        
+        private const string TemporaryProjectTemplate = @"<Project Sdk=""Microsoft.NET.Sdk"">
+  <PropertyGroup>
```

Comment:
Creating xml by formatting a string with holes is error prone. It will break if anything ever need escaping. Construct an XDocument and save it. The code can still read like a document: https://github.com/dotnet/xliff-tasks/blob/e1c2dadd49f151caa8d85516898375b31e7e1f38/src/XliffTasks/Model/XlfDocument.cs#L44

Reply:
Fixed, using XElement, reminds me of jquery good old days